File: homy/integrations/glances/widgets/system/system.py
"""Glances system monitor widget — bars, processes, network."""
import logging

from flask import jsonify, request

from homy.cache import cache_key, should_bypass_cache, widget_cache
from homy.integration_widget_util import get_integration_for_widget, get_widget_config
from homy import integration_widget_fetch as iwf

logger = logging.getLogger(__name__)

WIDGET = {
    'type': 'glances_system',
    'name': 'Glances System',
    'integration_types': ['glances'],
    'default_size_x': 6,
    'default_size_y': 5,
    'icon': 'cpu',
    'config_schema': {
        'integration_id': {'type': 'text', 'label': 'Integration ID', 'default': ''},
        'show_processes': {
            'type': 'select',
            'label': 'Prozesse anzeigen',
            'options': ['Ja', 'Nein'],
            'default': 'Ja',
        },
        'show_network': {
            'type': 'select',
            'label': 'Netzwerk anzeigen',
            'options': ['Ja', 'Nein'],
            'default': 'Nein',
        },
        'show_swap': {
            'type': 'select',
            'label': 'Swap anzeigen',
            'options': ['Ja', 'Nein'],
            'default': 'Nein',
        },
        'max_processes': {
            'type': 'select',
            'label': 'Max. Prozesse',
            'options': ['3', '5', '8'],
            'default': '5',
        },
    },
}

CACHE_NS = 'glances_system'
CACHE_TTL = 20


def register(app):
    @app.route('/api/glances_system/data', methods=['GET'], endpoint='widget_glances_system_data')
    def glances_system_data():
        widget_id = request.args.get('widget_id')
        if not widget_id:
            return jsonify({'error': 'Bad Request', 'message': 'widget_id required'}), 400

        wcfg = get_widget_config(widget_id)
        if wcfg is None:
            return jsonify({'error': 'Not Found'}), 404

        integration, config, err, code = get_integration_for_widget(widget_id, ['glances'])
        if err:
            return jsonify({'configured': False, 'online': False, 'message': err}), code

        ck = cache_key(CACHE_NS, widget_id)
        if not should_bypass_cache():
            cached = widget_cache.get(ck)
            if cached is not None:
                return jsonify(cached)

        try:
            result = iwf.fetch_glances_extended(config)
            try:
                max_proc = max(1, min(int(wcfg.get('max_processes', 5) or 5), 8))
            except (TypeError, ValueError):
                max_proc = 5
            result['processes'] = result.get('processes', [])[:max_proc]
            result['show_processes'] = wcfg.get('show_processes', 'Ja') == 'Ja'
            result['show_network'] = wcfg.get('show_network', 'Nein') == 'Ja'
            result['show_swap'] = wcfg.get('show_swap', 'Nein') == 'Ja'
            result['configured'] = True
            result['online'] = True
            widget_cache.set(ck, result, ttl=CACHE_TTL)
            return jsonify(result)
        except Exception as exc:
            logger.warning('glances_system failed: %s', exc)
            return jsonify({'configured': True, 'online': False, 'message': str(exc)}), 200

File: homy/integrations/glances/widgets/system/system.js
window.WidgetRegistry.register('glances_system', {
    _bar(label, pct, colorClass) {
        const p = Math.min(100, Math.max(0, pct ?? 0));
        const cls = p >= 90 ? 'gl-bar-danger' : p >= 70 ? 'gl-bar-warn' : (colorClass || '');
        return `<div class="gl-bar-item">
            <div class="gl-bar-label"><span>${label}</span><span>${p}%</span></div>
            <div class="gl-bar-track"><div class="gl-bar-fill ${cls}" style="width:${p}%"></div></div>
        </div>`;
    },

    _fmtMb(mb) {
        if (mb == null) return '—';
        if (mb >= 1024) return (mb / 1024).toFixed(1) + ' GB';
        return mb.toFixed(0) + ' MB';
    },

    async render(container, widgetData, config) {
        container.innerHTML = '<div class="widget-loading"><div class="spinner"></div></div>';
        try {
            const data = await API.request(`/api/glances_system/data?widget_id=${widgetData.id}`);
            container.innerHTML = '';
            if (!data.online) {
                container.innerHTML = `<div class="widget-error text-center" style="padding:12px;">${data.message || 'Not available'}</div>`;
                return;
            }

            const bars = [
                this._bar('CPU', data.cpu_percent, 'gl-bar-cpu'),
                this._bar('RAM', data.mem_percent, 'gl-bar-mem'),
                data.disk_percent != null ? this._bar(`Disk (${data.disk_mount || '/'})`, data.disk_percent, 'gl-bar-disk') : '',
                data.show_swap && data.swap_percent > 0 ? this._bar('Swap', data.swap_percent, 'gl-bar-swap') : '',
            ].filter(Boolean).join('');

            let netHtml = '';
            if (data.show_network) {
                netHtml = `<div class="gl-net">
                    <span class="gl-net-item"><i data-lucide="arrow-down" class="gl-net-icon gl-net-down"></i>${this._fmtMb(data.net_rx_mb)}</span>
                    <span class="gl-net-item"><i data-lucide="arrow-up" class="gl-net-icon gl-net-up"></i>${this._fmtMb(data.net_tx_mb)}</span>
                </div>`;
            }

            let procHtml = '';
            if (data.show_processes && data.processes?.length) {
                procHtml = `<div class="gl-proc-section">
                    <div class="gl-section-title">Prozesse</div>
                    <div class="gl-proc-list">
                        ${data.processes.map(p => `
                            <div class="gl-proc-row">
                                <span class="gl-proc-name" title="${p.name}">${p.name}</span>
                                <span class="gl-proc-stat">CPU ${p.cpu}%</span>
                                <span class="gl-proc-stat">MEM ${p.mem}%</span>
                            </div>
                        `).join('')}
                    </div>
                </div>`;
            }

            container.innerHTML = `<div class="gl-container">
                <div class="gl-bars">${bars}</div>
                ${netHtml}
                ${procHtml}
            </div>`;
            window.refreshLucideIcons?.(container);
        } catch (err) {
            container.innerHTML = `<div class="widget-error"><i data-lucide="alert-triangle"></i><span>${err.message}</span></div>`;
            window.refreshLucideIcons?.(container);
        }
    },
});
